### ISSUE TITLE
Compilation hangs when sorting

### ISSUE THREAD

**Issue by unrealwill** [wsmoses/Enzyme#127]
Hello, 

I'm trying to sort vector of integers inside the function I want to differentiate.
It should be a no operation in the backward pass. 
But the compilation hangs.

Can you advise for the proper way to tell enzyme to just ignore some variables or function ?
Thanks

```
#include <stdio.h>
#include <iostream>
#include <stdlib.h>
#include <random>
#include <math.h>
#include <vector>
#include <algorithm>

using namespace std;

double foo( double* __restrict__ parts,int n, int* cellId)
 {
    vector< int > sorted(n);
    for( int i = 0 ; i < n ; i++)
    {
        sorted[i] = cellId[i];
    }

    sort( sorted.begin(),sorted.end());

    double out = parts[0];

    return out;
}


int enzyme_dup;
int enzyme_out;
int enzyme_const;

typedef double (*f_ptr)(double *,int,int*);

extern double __enzyme_autodiff(f_ptr,
    int, double *, double *,
    int, int,
    int, int*);


int main() {

    srand(42);

    std::mt19937 e2(42);
    std::uniform_real_distribution<> dist(0, 10);
    int n = 100000;
    int d = 3;
    double* x = new double[n*d];
    double* d_x = new double[n*d];
    for( int i = 0 ; i < n*d ; i++)
    {
        x[i] = dist(e2);
        d_x[i] = 0.0;
    }

    int * cellId = new int[n];
    for( int i = 0 ; i < n ; i++)
    {
        cellId[i] = i;
    }

    printf("before autodiff\n");
    __enzyme_autodiff(foo,
        enzyme_dup, x, d_x,
        enzyme_const, n,
        enzyme_const, cellId);


    //printf("%f \n", y);
    for( int i = 0 ; i < 100 ; i++)
    {
    printf("dx[%d] = %f\n",i, d_x[i]);
    }

}
```
Compilation command : 
`clang test2.cpp -lstdc++ -lm -Xclang -load -Xclang /usr/local/lib/ClangEnzyme-7.so -O2 -fno-vectorize -fno-slp-vectorize -fno-unroll-loops -o test2
`

**Comment by dan-zheng** [comment]
> I'm trying to sort vector of integers inside the function I want to differentiate.
> It should be a no operation in the backward pass.

Note: in general, I believe the derivative of `sort` is a permutation, not a no-op. It involves shuffling derivative values.

<details>
<summary>Differentiable sort example from Swift</summary>
<p>

```swift
import _Differentiation

// Workaround because `sort` is defined as a default implementation on `Collection` protocols, not on `Array`.
func _sort<T: RandomAccessCollection & MutableCollection>(_ x: inout T)
where T.Element: Comparable {
  x.sort()
}

// Stub defining `sort` on `Array` via the `Collection` protocol default implementation.
extension Array where Element: Comparable {
  mutating func sort() {
    _sort(&self)
  }
}

// Register a derivative for `Array.sort`.
extension Array where Element: Comparable & Differentiable {
  @derivative(of: sort)
  mutating func vjpSort() -> (
    value: (), pullback: (inout TangentVector) -> Void
  ) {
    // Original computation: sort while obtaining index permutation.
    // https://stackoverflow.com/a/17554343/5365899
    let sortedIndices = indices.sorted(by: { i, j in self[i] < self[j] })
    var sorted = self
    for i in indices {
      sorted[i] = self[sortedIndices[i]]
    }
    self = sorted

    // Pullback: apply sorted permutation to tangent array.
    func pullback(_ v: inout TangentVector) {
      var sortedBase = v.base
      for i in sortedBase.indices {
        sortedBase[i] = v.base[sortedIndices[i]]
      }
      v.base = sortedBase
    }
    return ((), pullback)
  }
}

// Usage example.

let array: [Float] = (0..<10).map(Float.init).shuffled()
let pb = pullback(at: array) { array -> [Float] in
  var tmp = array
  tmp.sort() // differentiable sort invocation
  return tmp
}
let v = (0..<10).map(Float.init)
print("   Array:", array)
print("Gradient:", pb(.init(v)))
//    Array: [5.0, 6.0, 2.0, 9.0, 0.0, 8.0, 3.0, 1.0, 4.0, 7.0]
// Gradient: [4.0, 7.0, 2.0, 6.0, 8.0, 0.0, 1.0, 9.0, 5.0, 3.0]
```

</p>
</details>

**Comment by unrealwill** [comment]
I am sorting integers. They should be inactive.

**Comment by wsmoses** [comment]
Pardon the delay (just got to computer), will look into this after your other bug report.

Yes this should be inactive for many reasons including the type and also the fact it is not used.

My guess as to what is occurring here is that Enzyme's Type Analysis is being too aggressive and attempting to assess types down a recursive call chain (which will eventually terminate due to limits on how big it is allowed to go).

**Comment by wsmoses** [comment]
Ok I've confirmed this indeed to be the case. Specifying a smaller limit via `-mllvm -enzyme-max-type-offset=20` if you call with clang or `-enzyme-max-type-offset=20` if you call with opt remedies the long compile time (the default is 500, which is indeed likely way larger than it should be).

On my machine, however, all of the levels of introsort/etc that it analyzes/differentiates through provides a similar recursive issue (this time err'ing with Enzyme's cache analysis stack overflowing). I'm now looking into the second piece, but thought I'd provide an update regardless.

**Comment by wsmoses** [comment]
Also FWIW, answering your original question -- yes Enzyme does have a way to mark functions as inactive by definition (by specifying their LLVM attribute to include `enzyme_inactive`). This is currently used by the Julia frontend (and perhaps Rust, in progress), though we haven't added a binding for it to clang.

**Comment by unrealwill** [comment]
Thanks for the quick response :) (even more so on a sunday :)
I don't quite know the procedure to add some LLVM attributes in a c++ code (maybe some pragma ...)

**Comment by wsmoses** [comment]
I just added a quick and dirty "this is inactive" call mechanism both in #129 and also my dev branch.

Calling Enzyme as follows:
```
~/git/Enzyme/build/bin/clang sorter.cpp -lstdc++ -lm -Xclang -load -Xclang Enzyme/ClangEnzyme-7.so -O2 -o test2 -fno-vectorize -fno-slp-vectorize -fno-unroll-loops -fno-exceptions -mllvm -enzyme-max-type-offset=20 
```
with the following modification will no longer result in a compiler crash. I'm going to leave this issue open regardless for me to have the issue fixed without manual intervention, but hopefully this unblocks you.

```
void __enzyme_call_inactive(void*, ...);

double foo( double* __restrict__ parts,int n, int* cellId)
 {
    vector< int > sorted(n);
    for( int i = 0 ; i < n ; i++)
    {
        sorted[i] = cellId[i];
    }

    //sort( sorted.begin(),sorted.end());
    __enzyme_call_inactive((void*)std::sort<typeof(sorted.begin())>, sorted.begin(), sorted.end());

    double out = parts[0];

    return out;
}
```

**Comment by unrealwill** [comment]
Thanks, but now the Option Parser can't seem to parse the -mllvm -enzyme-max-type-offset=20

```
clang test2.cpp -lstdc++ -lm -Xclang -load -Xclang ~/Enzyme/enzyme/build/Enzyme/ClangEnzyme-7.so -O2 -o test2 -fno-vectorize -fno-slp-vectorize -fno-unroll-loops -fno-exceptions -mllvm -enzyme-max-type-offset=20 
clang (LLVM option parsing): for the -enzyme-max-type-offset option: Cannot find option named '20'!
```

I have done 
`git checkout fortranrb`
and remove then rebuild the build directories (llvm and enzyme) following the installation procedure.

**Comment by wsmoses** [comment]
It's likely that you may have needed to git fetch origin or the likes during checkout. Regardless, #120 has been merged into master which has the cache optimization that you desire and the following commit in master (https://github.com/wsmoses/Enzyme/commit/21bb9d01b8de0210008a0d3957c770f4af4e880b) should ensure the command line flag functions as expected.

So in short, repull master and use that and let us know if it still errs?

**Comment by unrealwill** [comment]
Just did that, and it works fine :) :+1: 
I'll let you close the issue when you see fit as it seems you wanted for enzyme to be able to do so without manual intervention or extra-flags.

**Comment by wsmoses** [comment]
The original sort code without special handling now "works" though type analysis is still quite slow (3m50s on my system without special speedup flags).

That said the odd segfault during compilation error is now gone so I will close the issue for now (and look into speeding up type analysis).
